internal/test/envtest/environment.go

```diff
@@ -0,0 +1,279 @@
+	controlplanev1beta1 "github.com/k0sproject/k0smotron/api/controlplane/v1beta1"
+	infrastructurev1beta1 "github.com/k0sproject/k0smotron/api/infrastructure/v1beta1"
+	k0smotronv1beta1 "github.com/k0sproject/k0smotron/api/k0smotron.io/v1beta1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
+)
+
+var (
+	cacheSyncBackoff = wait.Backoff{
+		Duration: 100 * time.Millisecond,
+		Factor:   1.5,
+		Steps:    8,
+		Jitter:   0.4,
+	}
+)
+
+// Environment encapsulates a Kubernetes local test environment.
+type Environment struct {
+	manager.Manager
+	client.Client
+	Config *rest.Config
+	env    *envtest.Environment
+	cancel context.CancelFunc
+}
+
+func init() {
+	logger := klog.Background()
+	// Use klog as the internal logger for this envtest environment.
+	log.SetLogger(logger)
+	// Additionally force all controllers to use the Ginkgo logger.
+	ctrl.SetLogger(logger)
+	// Add logger for ginkgo.
+	klog.SetOutput(ginkgo.GinkgoWriter)
+
+	utilruntime.Must(apiextensionsv1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(clusterv1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(k0smotronv1beta1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(bootstrapv1beta1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(controlplanev1beta1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(infrastructurev1beta1.AddToScheme(scheme.Scheme))
+}
+
+func newEnvironment() *Environment {
+	_, filename, _, _ := goruntime.Caller(0)
+	root := path.Join(path.Dir(filename), "..", "..", "..")
+
+	capiCoreCrdsPath := ""
+	if capiCoreCrdsPath = getFilePathToCAPICoreCRDs(); capiCoreCrdsPath == "" {
+		panic(fmt.Errorf("failed to retrieve cluster-api core crds path"))
+	}
+
+	crdPaths := []string{
+		capiCoreCrdsPath,
+		filepath.Join(root, "config", "clusterapi", "bootstrap", "bases"),
+		filepath.Join(root, "config", "clusterapi", "controlplane", "bases"),
+		filepath.Join(root, "config", "clusterapi", "infrastructure", "bases"),
+		filepath.Join(root, "config", "clusterapi", "k0smotron.io", "bases"),
+	}
+
+	env := &envtest.Environment{
+		ErrorIfCRDPathMissing: true,
+		CRDDirectoryPaths:     crdPaths,
+		CRDs: []*apiextensionsv1.CustomResourceDefinition{
+			builder.GenericInfrastructureMachineCRD.DeepCopy(),
+			builder.GenericInfrastructureMachineTemplateCRD.DeepCopy(),
+		},
+	}
+
+	if _, err := env.Start(); err != nil {
+		panic(err)
+	}
+
+	options := manager.Options{
+		Scheme: scheme.Scheme,
+		Metrics: metricsserver.Options{
+			BindAddress: "0",
+		},
+		Client: client.Options{
+			Cache: &client.CacheOptions{
+				DisableFor: []client.Object{
```

Comment:
If caching is not disabled in these cases, testing might encounter errors due to delays in cache synchronization, as the cache may not immediately reflect the current state in the API server